File: 0101_easy_hot_symmetric-tree/solution.java
import java.util.Deque;
import java.util.LinkedList;

/**
 * Definition for a binary tree node.
 */

 class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
 }

class Solution {

    public boolean isSymmetric2(TreeNode root) {
        Deque<TreeNode> queue = new LinkedList<>();

        queue.push(root.left);
        queue.push(root.right);

        while (!queue.isEmpty()) {
            TreeNode a = queue.pop();
            TreeNode b = queue.pop();

            if(a == null && b == null) {
                continue;
            }

            if(a == null || b == null) {
                return false;
            }

            if(a.val != b.val) {
                return false;
            }

            queue.push(a.left);
            queue.push(b.right);
            queue.push(a.right);
            queue.push(b.left);

        }

        return true;
    }

    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }

        return isSymmetric(root.left, root.right);
    }

    private boolean isSymmetric(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }

        if (left == null || right == null) {
            return false;
        }

        return (left.val == right.val && 
             isSymmetric(left.left, right.right) &&
             isSymmetric(left.right, right.left));
    }
}

File: 0101_easy_hot_symmetric-tree/solution.py
# Definition for a binary tree node.
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric2(self, root: Optional[TreeNode]) -> bool:

        if not root:
            return True

        queue = deque()

        queue.append(root.left)
        queue.append(root.right)

        while queue:
            a = queue.popleft()
            b = queue.popleft()

            if a is None and b is None:
                continue

            if a is None or b is None:
                return False

            if a.val != b.val:
                return False
            
            queue.append(a.left)
            queue.append(b.right)
            queue.append(a.right)
            queue.append(b.left)
        
        return True



    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.isSame(root.left, root.right)

    def isSame(self, left, right):
        if left is None and right is None:
            return True
        
        if left is None and right is not None:
            return False

        if left is not None and right is None:
            return False

        return left.val == right.val and self.isSame(left.left, right.right) and self.isSame(left.right, right.left)